torch/nested/_internal/ops.py

```diff
@@ -255,19 +267,461 @@ def jagged_binary_pointwise(func, *args, **kwargs):
+    # is not: batch, {seq_len}, num_heads, head_dim then we need
+    # to call contiguous
+    if not q_t.is_contiguous() and not _is_safe_to_get_storage_as_tensor(q_t):
+        q_t = q_t.contiguous()
+    if not k_t.is_contiguous() and not _is_safe_to_get_storage_as_tensor(k_t):
+        k_t = k_t.contiguous()
+    if not v_t.is_contiguous() and not _is_safe_to_get_storage_as_tensor(v_t):
+        v_t = v_t.contiguous()
+
+    query_buffer_reshaped = _view_as_dense(q_t, Nnz_q, num_heads, head_dim_qk)
+    key_buffer_reshaped = _view_as_dense(k_t, Nnz_kv, num_heads, head_dim_qk)
+    value_buffer_reshaped = _view_as_dense(v_t, Nnz_kv, num_heads, head_dim_v)
+
+    output_shape = q_t._size
+    if head_dim_v != head_dim_qk:
+        output_shape[-1] = head_dim_v
+
+    return (
+        query_buffer_reshaped,
+        key_buffer_reshaped,
+        value_buffer_reshaped,
+        cumulative_sequence_length_q,
+        cumulative_sequence_length_kv,
+        max_seqlen_batch_q,
+        max_seqlen_batch_kv,
+        output_shape,
+    )
+
+
+def jagged_scaled_dot_product_attention(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attn_mask: Optional[torch.Tensor] = None,
+    dropout_p=0.0,
+    is_causal=False,
+    scale=None,
+):
+    _validate_sdpa_input(query, key, value, attn_mask, dropout_p, is_causal, scale)
+    compute_logsumexp = query.requires_grad or key.requires_grad or value.requires_grad
+
+    backend_choice = _select_sdp_backend(
+        query, key, value, attn_mask, dropout_p, is_causal
+    )
+
+    if backend_choice == SDPBackend.FLASH_ATTENTION:
+        raise RuntimeError(
+            "Dispatcher for nested tensors with jagged layout cannot run Flash Attention v2 just yet."
+        )
+    elif backend_choice == SDPBackend.EFFICIENT_ATTENTION:
+        (
+            query_reshaped,
+            key_reshaped,
+            value_reshaped,
+            cumulative_sequence_length_q,
+            cumulative_sequence_length_kv,
+            max_seqlen_batch_q,
+            _,
+            output_shape,
+        ) = _sdpa_nested_preprocessing(query, key, value)
+        (
+            attention,
+            log_sumexp,
+            seed,
+            offset,
+            max_seqlen_q,
+            max_seqlen_batch_kv,
+        ) = torch.ops.aten._efficient_attention_forward(
+            query_reshaped.unsqueeze(0),
+            key_reshaped.unsqueeze(0),
+            value_reshaped.unsqueeze(0),
+            None,
+            cumulative_sequence_length_q,
+            cumulative_sequence_length_kv,
+            max_seqlen_batch_q,
+            dropout_p,
+            int(is_causal),
+            compute_logsumexp,
+            scale,
+        )
+
+        # Reshape output to convert nnz to batch_size and seq_len
+        return attention.view(-1).reshape(*output_shape).transpose(1, 2)
+    elif backend_choice == SDPBackend.MATH:
+        _can_use_math_sdpa_jagged(query, key, value)
```

Comment:
why call `_can_use_math_sdpa_jagged()` here instead of in backend selection where the other `can_use_*` helpers are called?

Reply:
there's no `can_use_math`, so instead I had to come up with one with the extra checks jagged requires for this particular case.

Reply:
oh, misunderstood the question, I'll move it there, yes, it makes more sense.

Reply:
moved it there